src/site_apportionment/site_apportionment.py

```diff
@@ -0,0 +1,219 @@
+import pandas as pd
+from pandas.api.types import is_numeric_dtype
+import os
+import logging
+
+SitesApportionmentLogger = logging.getLogger(__name__)
+
+# Colunm names redefined for convenience
+ref = "reference"
+ins = "instance"
+period = "period"
+form = "formtype"
+postcode = "postcodes_harmonised"
+percent = "602"
+product = "201"
+pg_num = "pg_numeric"
+civdef = "200"
+
+# Long and short form codes
+short_code = "0006"
+long_code = "0001"
+
+
+def count_unique_codes_in_col(df: pd.DataFrame, code: str) -> pd.DataFrame:
+    """Calculates the number of unique non-empty codes in a column.
+
+    Args:
+        df (pd.DataFrame): A dataframe containing all data
+        code (str): Name of the column containing codes
+
+    Returns:
+        (pd.DataFrame): A copy of originl dataframe with an additional column
+        called the same as code with suffix "_count" countaining the number of
+        unique non-empty codes
+    """
+
+    dfa = df.copy()
```

Comment:
I am personally unsure about the use of `df.copy()` at the start of this (and other) functions; I think it's unnecessary. 

As the the subsequent operations do not modify the original dataframe df, there's no need to create a copy. Removing this line could improve performance, especially as our dataframe is quite large.

Certainly happy to be proven wrong about this because I know other coders have used this throughout the codebase.

Reply:
I tend to be unsure about this- it seems to be necessary when filtering. Let's come back later and check with the debugger.